File: LeetCode/DP/JS/Leetcode873-length-of-longest-fibonacci-subsequence.js
var lenLongestFibSubseq = function (arr) {
    let [n, res, mp] = [arr.length, 0, new Map()];
    let f = new Array(n).fill(0).map(() => new Array(n).fill(0));
    for (let i = 0; i < n; i++) {
        mp.set(arr[i], i);
    }
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < i; j++) {
            const x = arr[i] - arr[j];
            f[i][j] = 2;
            if (x < arr[j] && mp.has(x)) {
                const k = mp.get(x);
                f[i][j] = Math.max(f[i][j], f[j][k] + 1);
            }
            res = Math.max(res, f[i][j]);
        }
    }
    if (res < 3) return 0;
    return res;
};

File: LeetCode/DP/Golang/Leetcode873-length-of-longest-fibonacci-subsequence.go
package sbgo

func lenLongestFibSubseq(arr []int) int {
	n, res, mp := len(arr), 0, make(map[int]int)
	for i := 0; i < n; i++ {
		mp[arr[i]] = i
	}
	f := make([][]int, n)
	for i := 0; i < n; i++ {
		f[i] = make([]int, n)
	}
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			x := arr[i] - arr[j]
			f[i][j] = 2
			if k, ok := mp[x]; x < arr[j] && ok {
				f[i][j] = max(f[i][j], f[j][k]+1)
			}
			res = max(res, f[i][j])
		}
	}
	if res < 3 {
		return 0
	}
	return res
}

func max(i, j int) int {
	if i < j {
		return j
	}
	return i
}
